openpgp/keys.go

```diff
@@ -194,7 +204,15 @@ func (el EntityList) KeysById(id uint64) (keys []Key) {
 
 		for _, subKey := range e.Subkeys {
 			if subKey.PublicKey.KeyId == id {
-				keys = append(keys, Key{e, subKey.PublicKey, subKey.PrivateKey, subKey.Sig})
+
+				// If there's both a a revocation and a sig, then take the
+				// revocation. Otherwise, we can proceed with the sig.
+				sig := subKey.Revocation
+				if sig == nil {
+					sig = subKey.Sig
+				}
+
+				keys = append(keys, Key{e, subKey.PublicKey, subKey.PrivateKey, sig})
 			}
 		}
 	}
```

Comment:
A note for KeysByIdUsage below: It looks like this function is also going to have problems with implicitly encryptable ElGamal keys. Maybe we need to factor out some shared function for all this "is this an encryption key" logic?